### ISSUE TITLE
5.15-rc6 vs 5.10: iou-wrk-# spawned too much; iou-sqp-# hang; linked timeout returen -EINVAL

### ISSUE THREAD

**Issue by beldzhang** [axboe/liburing#460]
We are convert our storage protocol to io_uring, on kernel 5.10 it worked perfectly, when try run on kernel 5.15, there are multi issues, I'm not sure there are related so I put them together.

The protocol has typical structure:
loop:
  1) wait header
  2) handle requests: read-and-send or recv-and-write

we use liburing 2.1, compared kernel 5.10 and 5.15-rc6:

issue 1:
when SQPOLL not set, lots of iou-wrk-#### spawned ( >100 ) when connections >= 2, 
I *think* it could be limited by io_uring_register_iowq_max_workers()
It happens when the disk read/write via splice way, if worked by buffered way, is ok.

issue 2:
when SQPOLL is set, after program exit, iou-sqp-#### hung up at state D (disk sleep)
the program can not be KILL, only do a system reboot

issue 3:
every network recv/send was linked with a timeout SQE, sometimes this SQE linked to header recving returned -EINVAL by cqe.res, this never occurred on kernel 5.10.
for now I just treat it like ECANCELED and looks ok. 

and other minor issues:
a) io_uring_submit() return 0 sometimes, is this an error?, we just try accept this as good, and looks everything good.

b) when timeout was set in io_uring_wait_cqe_timeout(), io_uring_wait_cqes(), then this calls can not be interrupted by signal either SIGINT or SIGTERM, this was found from liburing 0.7

more details can be provided if required, and any suggestion or code testing are welcome.

Thanks a lot for your hard working.

**Comment by isilence** [comment]
> issue 1: when SQPOLL not set, lots of iou-wrk-#### spawned ( >100 ) when connections >= 2, I _think_ it could be limited by io_uring_register_iowq_max_workers() It happens when the disk read/write via splice way, if worked by buffered way, is ok.

What is your rlimit value for RLIMIT_NPROC?

**Comment by isilence** [comment]
> issue 2: when SQPOLL is set, after program exit, iou-sqp-#### hung up at state D (disk sleep) the program can not be KILL, only do a system reboot

dmesg output may be useful. Also, can you get us a stack trace of that hung iou-sqp task?
```sh
cat /proc/<task-pid>/stack
```

May need to run several times to get anything meaningful if it's a live lock.

**Comment by isilence** [comment]
> issue 3: every network recv/send was linked with a timeout SQE, sometimes this SQE linked to header recving returned -EINVAL by cqe.res, this never occurred on kernel 5.10. for now I just treat it like ECANCELED and looks ok.

Is it SQPOLL? Do you share SQPOLL backends between multiple io_uring instances (i.e. `IORING_SETUP_ATTACH_WQ`)?
My guess it's either submitted a not linked IORING_OP_LINK_TIMEOUT, e.g. because of some SQPOLL restrictions. Or you have a IORING_OP_LINK_TIMEOUT linked to another IORING_OP_LINK_TIMEOUT.

Will you be able to test kernel patches?

**Comment by isilence** [comment]
reproduced _issue 3)), IOW we'll fix it soon

**Comment by isilence** [comment]
Found one problem relevant to __issue 3__, but it's not necessarily the one you have.
1) Do you call io_uring_register_iowq_max_workers() before issuing any request?
2) Do you call it from the thread/process that created the ring?
3) Do you call it from the thread/process that submits requests?

The thing is that `io_uring_register_iowq_max_workers()` is per process, and with some additional rules, we need to improve it a bit.

Also, can you tell your full name and email, so we can attach a reported-by tag to commits to give you some credit?
e.g. `Reported-by: Pavel Begunkov <my_email@example.com>`

**Comment by isilence** [comment]
The reproducer was pretty helpful thanks. Recapping:

> issue 1: when SQPOLL not set, lots of iou-wrk-#### spawned ( >100 ) when connections >= 2, I _think_ it could be limited by io_uring_register_iowq_max_workers() It happens when the disk read/write via splice way, if worked by buffered way, is ok.

There is a couple of new patches in the tree, I'd want you to try them. 

> issue 2: when SQPOLL is set, after program exit, iou-sqp-#### hung up at state D (disk sleep) the program can not be KILL, only do a system reboot

very likely it's because of __issue 3__

> issue 3: every network recv/send was linked with a timeout SQE, sometimes this SQE linked to header recving returned -EINVAL by cqe.res, this never occurred on kernel 5.10. for now I just treat it like ECANCELED and looks ok.

Fixed, patch is sent.

> and other minor issues: a) io_uring_submit() return 0 sometimes, is this an error?, we just try accept this as good, and looks everything good.

Can happen with SQPOLL and the userspace should be able to tolerate it, that's how liburing (not to mix with the kernel side) handles SQPOLL submissions. E.g. SQPOLL reaps some requests in the middle of submission and io_uring_submit() might return less than the number of SQE that had been in the queue.

> b) when timeout was set in io_uring_wait_cqe_timeout(), io_uring_wait_cqes(), then this calls can not be interrupted by signal either SIGINT or SIGTERM, this was found from liburing 0.7

confirmed not to happen in 5.15

**Comment by isilence** [comment]
> build and tested with this commit:
https://git.kernel.dk/cgit/linux-block/commit/?h=io_uring-5.15&id=46cb76b2f5ff39bf019bf7a072524fc7fe6deb01

Sorry, I'd usually give more instructions and links, but was waiting until Jens takes the patch.

> issue 1) fixed

perfect

> **issue 2) still, can be reproduced by server/cli_read** stack trace was the same

Yeah, rebase broke it. Jens will hopefully fix the branch today, but if not I create a new one, basically Jens' io_uring-5.15~1 + that patch.  Could you try it?
https://github.com/isilence/linux/tree/unprep

```sh
git clone https://github.com/isilence/linux.git
cd linux
git checkout origin/unprep
# and all the same further
```

> **issue 4) system load increasing slowly** after testing issue 1), load[1] increase 0.01 per second, until 2.00. before test it is ok. almost 0.0# checked by htop: (show thread, show kernel thread), can not indicate which one caused this. no any iou- wq left(checked by ps -A) dmesg: nothing output after system boot.

Can you tell if that's still the case after retesting?

**Comment by beldzhang** [comment]
Add some debug printk, found the max_workers bumped to RLIMIT_NPROC

change on 5.15-rc7
```
--- ../work/00-tmpfs/io-wq.c	2021-10-29 17:12:29.020248700 -0400
+++ fs/io-wq.c	2021-10-29 17:06:50.741514454 -0400
@@ -261,6 +261,9 @@
 	raw_spin_unlock(&wqe->lock);
 	atomic_inc(&acct->nr_running);
 	atomic_inc(&wqe->wq->worker_refs);
+
+	printk("uring-debug: io_wqe_create_worker(): %d vs %d\n", acct->nr_workers, acct->max_workers);
+
 	return create_io_worker(wqe->wq, wqe, acct->index);
 }
 
@@ -290,6 +293,7 @@
 	}
 	raw_spin_unlock(&wqe->lock);
 	if (do_create) {
+		printk("uring-debug: create_worker_cb(): %d vs %d\n", acct->nr_workers, acct->max_workers);
 		create_io_worker(wq, wqe, worker->create_index);
 	} else {
 		atomic_dec(&acct->nr_running);
```

dmesg (partial):
```
[  123.189722] uring-debug: io_wqe_create_worker(): 2 vs 2
[  123.190120] uring-debug: io_wqe_create_worker(): 2 vs 2
[  123.190199] uring-debug: io_wqe_create_worker(): 2 vs 2
[  125.565649] uring-debug: io_wqe_create_worker(): 1 vs 257437
[  125.612244] uring-debug: io_wqe_create_worker(): 1 vs 257437
[  125.612533] uring-debug: io_wqe_create_worker(): 2 vs 257437
[  125.628494] uring-debug: io_wqe_create_worker(): 3 vs 257437
[  125.930844] uring-debug: create_worker_cb(): 4 vs 257437
[  125.964480] uring-debug: create_worker_cb(): 5 vs 257437
[  125.967301] uring-debug: create_worker_cb(): 6 vs 257437
```

full dmesg output attached:
[dmesg-max_workers-bumped.txt](https://github.com/axboe/liburing/files/7444972/dmesg-max_workers-bumped.txt)

**Comment by isilence** [comment]
Can you make sure you call io_uring_register_iowq_max_workers(1,4) (or whatever arguments were) at least once for **each** ring? The limit is per task, but how the it's propagated is a different story. Just guessing how your app is written and want to check a hypothesis.

**Comment by beldzhang** [comment]
In my code, the `io_uring_register_iowq_max_wokers(1,2)` was just after the `io_uring_queue_init_params()`, they are wrapped to a single method of class.

after a reviews for the `epoll` to `io_uring` conversion, I think I'm **misunderstanding** of the "ring sharing":
I was create all the rings(call the above method) in a separated manage thread, and then pass each one to a working thread, so the ring is created in a **task**, and the queue loop is run in others.

then change the ring creation code to the begin of working thread, and tested with https://github.com/isilence/linux/commit/72c7f89c96fc7a39ff13109dd53271adf773d1fd

non-SQPOLL: it's good now

SQPOLL: iou-wrk-### still over created, but if I change to following sequence, it became good then ...
```
io_uring_queue_init_param(...)
if io_uring_register_iowq_max_workers(0,0) = 0 then begin
    io_uring_register_iowq_max_workers(1,2);
    io_uring_register_iowq_max_workers(0,0);  // ioq-wrk-### created as desired after this line added
end
```
BTW, my habit is set the resources usage to the minimal, (1,2) is the smallest value for the ring to work.

**Comment by beldzhang** [comment]
found a problem in io-wq.c: io_wq_max_workers(), do some patch, and **issue 1.1** is fixed.

before:
```
[  127.307054] uring-debug: io_wq_max_workers: acct=00000000b5415868, node=0, idx=0, new=4, prev=64
[  127.307060] uring-debug: io_wq_max_workers: acct=00000000c59638b2, node=0, idx=1, new=8, prev=257451
[  127.307061] uring-debug: io_wq_max_workers: acct=00000000a51d5e92, node=1, idx=0, new=64, prev=257451
[  127.307063] uring-debug: io_wq_max_workers: acct=000000005a29a9ed, node=1, idx=1, new=257451, prev=257451
```
after fix:
```
[   43.310441] uring-debug: io_wq_max_workers: acct=000000007a3a1926, node=0, idx=0, new=4, prev=64
[   43.310446] uring-debug: io_wq_max_workers: acct=000000005d1d6a6b, node=0, idx=1, new=8, prev=257451
[   43.310448] uring-debug: io_wq_max_workers: acct=00000000ad946a86, node=1, idx=0, new=4, prev=64
[   43.310450] uring-debug: io_wq_max_workers: acct=00000000c816e00d, node=1, idx=1, new=8, prev=257451
```
temporary patch i tested
```
diff --git a/fs/io-wq.c b/fs/io-wq.c
index 8fcac49b7b13..bddabbe105d3 100644
--- a/fs/io-wq.c
+++ b/fs/io-wq.c
@@ -1306,7 +1306,8 @@ int io_wq_cpu_affinity(struct io_wq *wq, cpumask_var_t mask)
  */
 int io_wq_max_workers(struct io_wq *wq, int *new_count)
 {
-	int i, node, prev = 0;
+	int i, node;//, prev = 0;
+	int prev[IO_WQ_ACCT_NR];
 
 	BUILD_BUG_ON((int) IO_WQ_ACCT_BOUND   != (int) IO_WQ_BOUND);
 	BUILD_BUG_ON((int) IO_WQ_ACCT_UNBOUND != (int) IO_WQ_UNBOUND);
@@ -1317,6 +1318,9 @@ int io_wq_max_workers(struct io_wq *wq, int *new_count)
 			new_count[i] = task_rlimit(current, RLIMIT_NPROC);
 	}
 
+	for (i = 0; i < IO_WQ_ACCT_NR; i++)
+		prev[i] = 0;
+
 	rcu_read_lock();
 	for_each_node(node) {
 		struct io_wqe *wqe = wq->wqes[node];
@@ -1325,13 +1329,18 @@ int io_wq_max_workers(struct io_wq *wq, int *new_count)
 		raw_spin_lock(&wqe->lock);
 		for (i = 0; i < IO_WQ_ACCT_NR; i++) {
 			acct = &wqe->acct[i];
-			prev = max_t(int, acct->max_workers, prev);
+			if (node == 0)
+				prev[i] = max_t(int, acct->max_workers, prev[i]);
 			if (new_count[i])
 				acct->max_workers = new_count[i];
-			new_count[i] = prev;
+			//printk("uring-debug: io_wq_max_workers: acct=%p, node=%d, idx=%d, new=%d, prev=%d\n", acct, node, i, new_count[i], prev[i]);
+			//new_count[i] = prev;
 		}
 		raw_spin_unlock(&wqe->lock);
 	}
+	for (i = 0; i < IO_WQ_ACCT_NR; i++)
+		new_count[i] = prev[i];
+
 	rcu_read_unlock();
 	return 0;
 }
```

**Comment by isilence** [comment]
Awesome, makes sense, slipped from my head that there are multiple nodes, wanted to kill it off but forgot.
Would be great if you send a patch to `io-uring@vger.kernel.org` (don't forget to add signed-off-by), though we can take over if you don't have time.

**Comment by beldzhang** [comment]
patch sent.

**Comment by beldzhang** [comment]
I think this need to be patched to 5.15.x also, shall I send to somewhere else?

**Comment by axboe** [comment]
It'll automatically go to 5.15.x, that's why I marked it with a fixes tag. You don't need to do anything else.

**Comment by beldzhang** [comment]
@isilence 
https://github.com/isilence/linux/commit/72c7f89c96fc7a39ff13109dd53271adf773d1fd still required as it fixed issue 1.2

**Comment by axboe** [comment]
Pavel, are you sending that one in for 5.16?
